### PR TITLE
Fix Annotate Target to support freevars(relay.zeros, relay.ones etc) of any size (including zero) 

### PR DESCRIPTION
**Pull request by codeislife99**
In the Torchvision 0.7 MaskRCNN model there are freevars(zeros, ones) of shape 0. Annotate target pass currently generates a compiler end (without a compiler begin) for all of such cases. This PR fixes that.

### CODE REVIEW

File: src/relay/transforms/annotate_target.cc [hunk 1]
```diff
@@ -77,7 +77,11 @@ class AnnotateTargetRewriter : public ExprRewriter {
         compiler_ends.push_back(call->args[0]);
       } else if (op_expr_to_target_.find(arg) != op_expr_to_target_.end()) {
         arg_target = op_expr_to_target_[arg];
-        compiler_ends.push_back(InsertAnnotation(arg, arg_target, make_end_op));
+        if (call && call->args.size() == 0) {
```
Comment (comaniac) [inline src/relay/transforms/annotate_target.cc:80]: Please add comments to indicate in which case we don't want `compiler_end`.

File: src/relay/transforms/annotate_target.cc [hunk 2]
```diff
@@ -120,7 +124,7 @@ class AnnotateTargetRewriter : public ExprRewriter {
      * \return An annotated and target-propagated relay expression.
      */
     Expr new_expr = expr;
-    if (op_expr_to_target_.find(expr) != op_expr_to_target_.end()) {
+    if (op_expr_to_target_.find(expr) != op_expr_to_target_.end() && FreeVars(expr).size() != 0) {
```
Comment (comaniac) [inline src/relay/transforms/annotate_target.cc:127]: Ditto. Please comment why we skip `FreeVars`.

**Review by comaniac — CHANGES_REQUESTED** [review]
Thanks for the efforts :) Please also add test cases to cover the new cases you worked.

**Comment by codeislife99** [comment]
Yes sorry, I didn't mean to open it for review. I am still working on it :)

File: src/relay/transforms/annotate_target.cc (hunk 1 above)
Comment (codeislife99) [reply src/relay/transforms/annotate_target.cc:81]: Done.

File: src/relay/transforms/annotate_target.cc (hunk 2 above)
Comment (codeislife99) [reply src/relay/transforms/annotate_target.cc:130]: Done and added test cases.

File: tests/python/relay/test_pass_annotate_target.py [hunk 3]
```diff
@@ -510,6 +510,188 @@ def after():
...
+
+    @tvm.ir.register_op_attr("sigmoid", "target." + target)
+    def sigmoid(attrs, args):  # pylint: disable=unused-variable
+        return True
+
+    @tvm.ir.register_op_attr("erf", "target." + target)
+    def erf(attrs, args):  # pylint: disable=unused-variable
+        return True
+
+    """Test that If-else nodes compiles correctly when surrounded by free variables"""
```
Comment (comaniac) [inline tests/python/relay/test_pass_annotate_target.py:528]: Move this line under `def test_if_free_vars_1():`

File: tests/python/relay/test_pass_annotate_target.py [hunk 4]
```diff
@@ -510,6 +510,188 @@ def after():
...
```
Comment (comaniac) [inline tests/python/relay/test_pass_annotate_target.py:595]: Ditto. Move to the function declaration.

Also from what is being tested in this function, it seems to me that the purpose of having this test is zero shape instead of If-else node IIUC. If so please update the docstring and the function name to reflect it.

File: tests/python/relay/test_pass_annotate_target.py [hunk 5]
```diff
@@ -510,6 +510,188 @@ def after():
...
+
+    result = transform.AnnotateTarget(target)(before())
+    expected = transform.InferType()(after())
+    assert tvm.ir.structural_equal(expected, result)
+
+
+def test_free_vars_zeros_1():
+    target = "test_free_vars_zeros_1"
+
+    """Test that free variables compile correctly on their own"""
```
Comment (comaniac) [inline tests/python/relay/test_pass_annotate_target.py:650]: Ditto. Also as we already have the above two tests, do we really need this one?

File: tests/python/relay/test_pass_annotate_target.py [hunk 6]
```diff
@@ -510,6 +510,188 @@ def after():
...
+
+    def before():
+
```
Comment (comaniac) [inline tests/python/relay/test_pass_annotate_target.py:653]: remove this line.

File: tests/python/relay/test_pass_annotate_target.py [hunk 7]
```diff
@@ -510,6 +510,188 @@ def after():
...
+        func = relay.Function([], relay.zeros(shape=(1, 32), dtype="float32"))
+        mod = tvm.IRModule.from_expr(func)
+
+        return mod
+
+    def after():
+
```
Comment (comaniac) [inline tests/python/relay/test_pass_annotate_target.py:660]: remove this line.

**Comment by codeislife99** [comment]
cc: @anijain2305 @comaniac

File: tests/python/relay/test_pass_annotate_target.py [hunk 8]
```diff
@@ -510,6 +510,188 @@ def after():
...
+        mod = tvm.IRModule.from_expr(func)
+
+        return mod
+
+    result = transform.AnnotateTarget(target)(before())
+    expected = transform.InferType()(after())
+    assert tvm.ir.structural_equal(expected, result)
+
+
+def test_free_vars_zeros_2():
```
Comment (comaniac) [inline tests/python/relay/test_pass_annotate_target.py:671]: Same question. Do we need this test?

**Review by comaniac — CHANGES_REQUESTED** [review]

File: src/relay/transforms/annotate_target.cc (hunk 1 above)
Comment (comaniac) [reply src/relay/transforms/annotate_target.cc:81]: Thanks. Would you be more specific like "If an argument is a call node and has no argument, then it should be tensor ops such as zeros, so we treat it as input vars."

Comment (codeislife99) [reply src/relay/transforms/annotate_target.cc:82]: Done.

File: tests/python/relay/test_pass_annotate_target.py (hunk 5 above)
Comment (codeislife99) [reply tests/python/relay/test_pass_annotate_target.py:581]: I removed the two unnecessary tests but I think we need one test for a fn node and one test for a regular one.

File: tests/python/relay/test_pass_annotate_target.py (hunk 8 above)
Comment (codeislife99) [reply tests/python/relay/test_pass_annotate_target.py]: Removed.

File: tests/python/relay/test_pass_annotate_target.py (hunk 7 above)
Comment (codeislife99) [reply tests/python/relay/test_pass_annotate_target.py]: Done.

File: tests/python/relay/test_pass_annotate_target.py (hunk 6 above)
Comment (codeislife99) [reply tests/python/relay/test_pass_annotate_target.py]: Done.

File: tests/python/relay/test_pass_annotate_target.py (hunk 4 above)
Comment (codeislife99) [reply tests/python/relay/test_pass_annotate_target.py]: Removed this test, not necessary.

File: tests/python/relay/test_pass_annotate_target.py (hunk 5 above)
Comment (anijain2305) [reply tests/python/relay/test_pass_annotate_target.py:581]: So, FunctionNode/IfNode go through a different codepath for inserting the final compiler_end. If the FunctionNode/IfNode has just one `zero` op, then it might lead to another compiler_end. Thats why, we need both tests.

Comment (anijain2305) [reply tests/python/relay/test_pass_annotate_target.py:581]: This is related to the FreeVars change in the source file.

**Review by comaniac — APPROVED** [review]
LGTM

**Comment by comaniac** [comment]
Thanks @codeislife99 @anijain2305
